django/db/backends/postgresql/base.py

```diff
@@ -221,35 +253,62 @@ def get_new_connection(self, conn_params):
         #   will set connection.isolation_level to ISOLATION_LEVEL_AUTOCOMMIT.
         options = self.settings_dict["OPTIONS"]
         try:
-            self.isolation_level = options["isolation_level"]
+            isolevel = options["isolation_level"]
         except KeyError:
-            self.isolation_level = connection.isolation_level
+            self.isolation_level = IsolationLevel.READ_COMMITTED
         else:
-            # Set the isolation level to the value from OPTIONS.
-            if self.isolation_level != connection.isolation_level:
-                connection.set_session(isolation_level=self.isolation_level)
-        # Register dummy loads() to avoid a round trip from psycopg2's decode
-        # to json.dumps() to json.loads(), when using a custom decoder in
-        # JSONField.
-        psycopg2.extras.register_default_jsonb(
-            conn_or_curs=connection, loads=lambda x: x
-        )
+            try:
+                self.isolation_level = IsolationLevel(isolevel)
+            except ValueError:
+                raise ImproperlyConfigured(
+                    "bad isolation_level: %s. Choose one of the "
+                    "'psycopg.IsolationLevel' values" % (options["isolation_level"],)
+                )
+            connection.isolation_level = self.isolation_level
+
+        connection.cursor_factory = Cursor
+
         return connection
 
     def ensure_timezone(self):
         if self.connection is None:
             return False
-        conn_timezone_name = self.connection.get_parameter_status("TimeZone")
+
+        conn_timezone_name = self.connection.info.parameter_status("TimeZone")
         timezone_name = self.timezone_name
         if timezone_name and conn_timezone_name != timezone_name:
             with self.connection.cursor() as cursor:
                 cursor.execute(self.ops.set_time_zone_sql(), [timezone_name])
             return True
         return False
 
+    @lru_cache(maxsize=2)
+    def get_adapters_template(self, use_tz):
```

Comment:
you can't put `@lru_cache` on a method - use a module level function.

see https://github.com/PyCQA/flake8-bugbear error B019 and https://www.youtube.com/watch?v=sVjtp6tGo0g

Reply:
Oh wow, this is makes total sense, but now we have a few things to fix. An `rg lru_cache` in Django shows quite a few hits. Like https://github.com/django/django/blob/main/django/apps/registry.py#L168 -- I mean `Apps` is probably maybe used as singleton but we might wanna fix those and add the flake8 warnings?